### DISCUSSION TITLE
Understanding Alignments

### DISCUSSION THREAD

**Discussion by MaveriQ** [explosion/spaCy#8367] [Help: Coding & Implementations] [answered]
I am trying to understand the alignment explanation given at : https://spacy.io/usage/linguistic-features#aligning-tokenization

In a minimal example I have the following

```
from spacy.training import Alignment
bert_tokens = ['we', 'can', "'", 't', 'go', '.']
spacy_tokens = ['We', 'ca', "n't", 'go', '.']
alignment = Alignment.from_strings(bert_tokens, spacy_tokens)
```
Now `alignment.x2y.dataXd` gives `[0, 1, 2, 2, 2, 3, 4]` and I can't make out how to map the three 2s in locations 2,3 and 4. I would appreciate any help.

**Comment by Pandalei97** [comment, marked as answer]
Hi,

In short, it's because there are two tokens in `spacy_tokens`(`ca` and `n't`) that refer to `can`.

You need to look at both `lengths` and `dataXd`.

`alignment.x2y.lengths` gives   `[1, 2, 1, 1, 1, 1]`
`alignment.x2y.dataXd ` gives  `[0, 1, 2, 2, 2, 3, 4]`

`alignment.x2y.lengths[1] == 2` means that `bert_tokens[1]` maps to two tokens in `spacy_tokens` and you will have two values in `alignment.x2y.dataXd ` (`dataXd[1]` and `dataXd[2]` in your case) that will correspond to `bert_tokens[1]`
